Custom Arrow Tips

behackl opened the pull request:
## List of Changes
- Refactor `ArrowTip` in `geometry.py` to be a (actually quite concrete) abstract base class so that different arrow tip shapes can be implemented by inheriting from it.
- Add `ArrowTriangleTip` (the current default tip), and two new tips `ArrowCircleTip`, `ArrowSquareTip`.
- Add keyword `tip_shape` to pass the custom shape to the `Arrow` constructor.



## Motivation
This is part of #272.

## Explanation for Changes
This PR changes the class hierarchy for arrow tips a little in order to allow the (easy) implementation of custom tips. Details are given in the List of Changes.

## Testing Status

As an example, rendering the `ArrowTest` scene in the Snippet

```python
import numpy as np

from manim import Scene, Arrow, ShowCreation, FadeOut,\
                  ArrowTriangleTip, ArrowSquareTip, ArrowCircleTip


class ArrowTest(Scene):
    def construct(self):
        a11 = Arrow(np.array([-2,3,0]), np.array([2, 3, 0]))
        a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
                    tip_shape=ArrowTriangleTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        a21 = Arrow(np.array([-2,1,0]), np.array([2, 1, 0]), tip_shape=ArrowSquareTip)
        a22 = Arrow(np.array([-2,0,0]), np.array([2, 0, 0]), tip_shape=ArrowSquareTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        a31 = Arrow(np.array([-2,-1,0]), np.array([2, -1, 0]), tip_shape=ArrowCircleTip)
        a32 = Arrow(np.array([-2,-2,0]), np.array([2, -2, 0]), tip_shape=ArrowCircleTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        self.play(ShowCreation(a11), ShowCreation(a12), ShowCreation(a21), 
                  ShowCreation(a22), ShowCreation(a31), ShowCreation(a32))
        self.wait(1)
        self.play(*[FadeOut(obj) for obj in self.mobjects])
```

yields

![ArrowTest](https://user-images.githubusercontent.com/11851593/91567107-e3f8c200-e944-11ea-94ba-4db76a1c549b.gif)



## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

huguesdevimeux commented:
Thanks for your contribution ! 
As this PR is still WIP; I'm marking it as draft.

leotrs commented:
WOWOWOW This is beautiful thanks so much! I'll review in a sec, but thank you!

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -809,3 +782,51 @@ class RoundedRectangle(Rectangle):
+        return get_norm(self.get_vector())
+
+
+class ArrowTriangleTip(ArrowTip, Triangle):
```
Comment: I'm usually leery of multiple inheritance, but seeing as these classes are (a) very simple and (b) not meant to be inherited from, then I _guess_ it's ok.

behackl replied:
I'm with you on being cautious regarding multiple inheritance. In this case, on the one hand, I believe it is actually semantically correct -- and on the other hand, I needed it for `ArrowCircleTip`: without having this class inherit from `Circle`, it would not populate `self.points`.

(But sure, I guess if it were necessary to remove multiple inheritance, a workaround could be found.)

leotrs replied:
Yeah, agreed. Perhaps we can just leave a line somewhere in the docs recommending that people not inherit from them.

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -809,3 +782,51 @@ class RoundedRectangle(Rectangle):
+        "start_angle": PI,
+    }
+
+    def get_base(self):
+        return self.point_from_proportion(0.5)
+
+    def get_tip_point(self):
+        return self.points[0]
+
+    def get_vector(self):
+        return self.get_tip_point() - self.get_base()
+
+    def get_angle(self):
+        return angle_of_vector(self.get_vector())
+
+    def get_length(self):
+        return get_norm(self.get_vector())
```
Suggestion: I'd rather all of these be properties

behackl replied:
Agreed. I changed `base`, `tip_point` and `vector` in 38638475195b9940561deb1f49404e61dafba19a. Actually, `get_length` is not used anywhere -- and trying to refactor it as a `length` property would require additional work as the very same `length` is also passed via a corresponding entry in the `CONFIG` dict of `ArrowTip` (from where it is subsequently used to construct the tips...). I've removed the superfluous `get_length` for now.

Finally, changing `get_angle` to an `angle`-property reveals another problem: a `ArrowCircleTip` inherits from `Circle` which inherits from `Arc` -- and an `Arc` also has an `angle` property. I've renamed the former to `tip_angle`, see a01dd7f1f963b44651136d8a4982b85029834e4e.

leotrs replied:
CONFIG dicts are going away soon, so we can perhaps revisit this when that happens. This is fine for now

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -809,3 +782,51 @@ class RoundedRectangle(Rectangle):
+        "length": DEFAULT_ARROW_TIP_LENGTH,
+        "start_angle": PI,
+    }
+
```
Suggestion: perhaps add
```python
def __init__(self):
    pass
```
Explicity is better than implicit and whatnot

behackl replied:
I fully agree. A more agressive suggestion is in bb1746d0b95510cf61e6b4bfe446ef1c55854738 (because I think it is better to avoid even the slightest chance of silent failures).

leotrs replied:
Oh I like your solution. I was about to say this should be an AbstractBaseClass, but that would have meant triple inheritance which I'm not going anywhere near.

behackl commented:
Thank you for this lightning-fast review! I was just typing that you don't need to stress yourself, as I don't know how soon I can add documentation and tests. 😄 

I'll look at your comments a bit later -- thanks again!

huguesdevimeux commented:
Speaking of tests, I think we should wait for #340 to be merged before adding new tests, so it will be easier.

behackl commented on manim/mobject/geometry.py:
```diff
@@ -597,12 +599,12 @@ class Arrow(Line):
         # Arrow.set_stroke is called?
         self.initial_stroke_width = self.stroke_width
-        self.add_tip()
+        self.add_tip(tip_shape=tip_shape)
```
79757610679448cc0560ccffc9578d0254b5133c

behackl replied:
I did this for the other arrow types as well; with 151d0a32 also start+end tip of double arrows and tips of curved arrows are customizable. :-)

behackl commented on manim/mobject/geometry.py:
```diff
@@ -597,12 +599,12 @@ class Arrow(Line):
     }
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, tip_shape=None, **kwargs):
```
79757610679448cc0560ccffc9578d0254b5133c

behackl replied:
I don't see a nice way to specify `ArrowTriangleTip` directly as the default value for `TipableVMobject.add_tip` (without some serious reordering I always get a `NameError` as `ArrowTriangleTip` is only defined later in the file -- and moving the implementation of the arrow tips before `TipableVMobject` is not possible because of `ArrowCircleTip`, which is a `Circle`, which is an `Arc`, which is a `TipableVMobject` again.)

Maybe this would work if the arrow tips were moved to a separate file -- but I think this might be overkill. I suggest simply handling the case of `tip_shape is None` (as it was before 7975761). Fixed via 249c2841.

leotrs replied:
Perfect, agreed on all accounts

behackl commented:
> One more suggestion: I think the current way of creating filled/hollow tips is a bit awkward. Why not add a filled class to each? In this way this code
> 
> ```python
> a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
>                     tip_shape=ArrowTriangleTip,
>                     tip_style={'fill_opacity': 0, 'stroke_width': 3})
> ```
> 
> becomes this
> 
> ```python
> a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
>                     tip_shape=ArrowTriangleFilledTip)
> ```

Great idea! It really was rather clunky before. I've added some convenience classes for filled/hollow tips in e8d076ca. :-)

leotrs commented:
Great great stuff. BTW, it was _extremely_ helpful of you to leave the commit link in each of your comments. Super easy to review!

behackl commented:
> Great great stuff. BTW, it was _extremely_ helpful of you to leave the commit link in each of your comments. Super easy to review!

Thank you for the kind words and your reviews! The thing with the commit hashes in the comments is what I like most when I have to review stuff, glad you like it as well. 😄 

I'll finish this here up by writing some docstrings, then the draft status of the PR can be removed again. As @huguesdevimeux suggested above, I'll wait with adding tests until #335 is ready (or take care of that in a follow-up PR). 👍

leotrs commented:
Sounds like a plan

behackl commented:
I have made all changes I wanted to make.

- 80a07d5 simplifies the `Arrow.scale` method, which scales arrows while leaving arrow tips unchanged.
- 7e2ddf8 adds a basic form of documentation. As nobody had any serious concerns on Discord, I also added some explicit doctests (which are not performed yet in the pipeline, but I would like to open a separate PR to enable just that).
- a112cd3 is just an application of `black` to `geometry.py`, and
- 78adf579 merges the current master in here (because I hoped to get rid of the format test failure).

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -807,22 +850,98 @@ def __init__(self, *args, **kwargs):
+    different default settings for styling arrow tips. These settings
+    (in particular `fill_opacity` and `stroke_width`) can also be
+    overridden manually.
+
```
Suggestion: Maybe add an Examples section that says something along the lines of "using ArrowFilledTip is a shorthand to using ArrowTip(...)`.

behackl replied:
I like that! 5528cc30a351ff57a5b06f3ab085620148cc2fe8

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -807,22 +850,98 @@ def __init__(self, *args, **kwargs):
+        Examples
+        --------
+
+        ::
```
Question: why not specify `.. code-block::` instead?

behackl replied:
AFAIK, it should not be necessary to do so, see https://www.sphinx-doc.org/en/1.5/markup/code.html -- but I admit, I did not try to actually generally the sphinx documentation.

leotrs replied:
I think you're right, it's not necessary, unless you want syntax highlighting. I think we want syntax highlighting :) Please use ` code-block:: python`

behackl replied:
I can add it, sure -- but the link I posted says that, as python is the default highlight language, it is sufficient to put `::` and still get syntax highlighting. :-)

(I'll try it later, currently I am not home.)

behackl replied:
https://manim.hackl.space/reference/manim.mobject.geometry.ArrowTip.html -- Syntax highlighting works indeed, but the formatting is off. I'll investigate and try to fix it.

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -807,22 +850,98 @@ def __init__(self, *args, **kwargs):
 
     @property
     def vector(self):
+        r"""The arrow tip vector spanning from base to tip.
```
Suggestion: "The vector with tail at the base of the arrow, and head at the tip of the arrow." (The word "span" has a precise meaning in linear algebra that may be confusing to some of our mathematically-adept readers (such as myself))

behackl replied:
While I don't really see the danger of interpreting this as the span in a vector space sense, especially given that it is used as a verb with specified start and end point, I don't mind rewording. I wanted to avoid tail and head though, so I propose a construction using "pointing" in 5528cc30a351ff57a5b06f3ab085620148cc2fe8.

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -795,6 +814,30 @@ def __init__(self, **kwargs):
+        ...
+        NotImplementedError: Has to be implemented in inheriting subclasses.
+
+    Instead, use it like this::
```
Suggestion: "Instead, use one of the pre-defined ones, or make a custom one like this:"

behackl replied:
Reworded in 5528cc30a351ff57a5b06f3ab085620148cc2fe8

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -795,6 +814,30 @@ def __init__(self, **kwargs):
+        ...             tip_shape=MyCustomArrowTip)
+        >>> isinstance(arr.tip, RegularPolygon)
+        True
+    
```
Suggestion: add a "See also" section that points to all the pre-defined classes

behackl replied:
List added in 5528cc30a351ff57a5b06f3ab085620148cc2fe8

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -613,6 +616,31 @@ def __init__(self, *args, **kwargs):
         self.set_stroke_width_from_length()
 
     def scale(self, factor, **kwargs):
```
Suggestion: I think this behavior should be the default, but it should be optional. Can we add a boolean that would allow the user to propagate the scale() call to the tips?

behackl replied:
Good call! cfb2cdaab5909b58859bc769cb914693b6bb951a

behackl commented:
Formatting inconsitencies are fixed in e0acdeb8.

Changing how tips are added to arrows also slightly changed the visuals of `DoubleArrow`, which is why the corresponding doctest has to be fixed in 48acf99c.
Before: 
![image](https://user-images.githubusercontent.com/11851593/91671430-eca8ee00-eb26-11ea-8836-cf68a7a8a758.png)

After: 
![image](https://user-images.githubusercontent.com/11851593/91671393-84f2a300-eb26-11ea-9df9-aef2101f1f9a.png)

(Notice that the tips were not equally large before, this is fixed now.)

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -666,12 +666,12 @@ def scale(self, factor, scale_tips=False, **kwargs):
         See Also
         --------
-        :meth:`~VMobject.scale`
+        :meth:`~.Mobject.scale`
```
Hmmm Line is a Vmobject, no? So it would be better to link to VMobject.scale instead. Why did you change it?

behackl replied:
There is no separate `VMobject.scale`, it is directly inherited from `Mobject.scale`, which caused the "See Also" block in the rendered HTML documentation just to include `scale()` without link (and in particular, without reference to `VMobject` or `Mobject`).

leotrs replied:
Ah good point

leotrs commented:
@behackl can you please solve the conflicts and make sure that the tests still pass? We just merged #335 which completely rehauled the test suite

behackl commented:
There was no actual conflict, I only needed to merge master locally. And of course, black still fails.

Unfortunately, the tests revealed another inconvenience: due to different architectures, numerics are slightly different between operating systems, which leads to numpy yielding `-0.0` in macos and `0.0` otherwise. 🙄

I'll come up with a fix for this, or replace the faulty doctests.

behackl commented:
- Fixed doctests via 5b6f3aa6
- And I also added a graphical unit test in 9f29fdfe -- this was very painless and straightforward, thanks to #335  :-)

Ready to be reviewed again!

leotrs commented on manim/mobject/geometry.py:
```diff
@@ -932,7 +932,7 @@ def base(self):
         ::
             >>> arrow = Arrow(np.array([0, 0, 0]), np.array([2, 0, 0]), buff=0)
-            >>> arrow.tip.base.round(2)
+            >>> arrow.tip.base.round(2) + 0.  # add 0. to avoid negative 0 in output
```
lol, I did not know about this hack

leotrs commented:
Can we get one more approval here? This is super ready @huguesdevimeux @Aathish04

Aathish04 reviewed (approved):
Thank you for the contributions!

huguesdevimeux reviewed (approved):
SGTM !

leotrs commented:
:partying_face:
